src/Network/Kademlia/Implementation.hs

```diff
@@ -22,115 +23,123 @@ import           Control.Concurrent.Chan     (Chan, newChan, readChan)
-
-          -- When receiving a RETURN_VALUE command, finish the lookup, then
-          -- cache the value in the closest peer that didn't return it and
-          -- finally return the value
-          checkSignal (Signal origin (RETURN_VALUE _ value)) = do
-                -- Abuse the known list for saving the peers that are *known* to
-                -- store the value
-                modify $ \s -> s { known = [origin] }
-
-                -- Finish the lookup, recording which nodes returned the value
-                finish
-
-                -- Store the value in the closest peer that didn't return the
-                -- value
-                known <- gets known
-                polled <- gets polled
-                let rest = polled \\ known
-                unless (null rest) $ do
-                    let cachePeer = peer . head $ sortByDistanceTo rest nid `usingConfig` config inst
-                    liftIO . send (handle inst) cachePeer . STORE nid $ value
-
-                -- Return the value
-                return . Just $ (value, origin)
-
-          -- When receiving a RETURN_NODES command, throw the nodes into the
-          -- lookup loop and continue the lookup
-          checkSignal (Signal _ (RETURN_NODES _ _ nodes)) =
-                continueLookup nodes sendS continue cancel
-          checkSignal _ = error "Fundamental error in unhandled query @lookup@"
-
-          -- Continuing always means waiting for the next signal
-          continue = waitForReply cancel checkSignal
-
-          -- Send a FIND_VALUE command, looking for the supplied id
-          sendS = sendSignal (FIND_VALUE nid)
-
-          -- As long as there still are pending requests, wait for the next one
-          finish = do
-                pending <- gets pending
-                unless (null pending) $ waitForReply (return ()) finishCheck
-
-          -- Record the nodes which return the value
-          finishCheck (Signal origin (RETURN_VALUE _ _)) = do
-                known <- gets known
-                modify $ \s -> s { known = origin:known }
-                finish
-          finishCheck _ = finish
+  where
+    go = startLookup (config inst) sendS cancel checkSignal
+
+    -- Return Nothing on lookup failure
+    cancel = pure Nothing
+
+    -- When receiving a RETURN_VALUE command, finish the lookup, then
+    -- cache the value in the closest peer that didn't return it and
+    -- finally return the value
+    checkSignal (Signal origin (RETURN_VALUE _ value)) = do
+        -- Abuse the known list for saving the peers that are *known* to
+        -- store the value
+        modify $ \s -> s { known = [peer origin] }
+
+        -- Finish the lookup, recording which nodes returned the value
+        finish
+
+        -- Store the value in the closest peer that didn't return the
+        -- value
+        known <- gets known
+        polled <- gets polled
+        rest <- resolveNodeIds inst (polled \\ known)
```

Comment:
I don't understand how  we can guarantee that nodes from `polled` are contained in buckets